kore/src/Kore/Builtin/Map.hs

```diff
@@ -561,3 +581,109 @@ unifyEquals unifyEqualsChildren first second = do
             normalizedOrBottom
                 :: Ac.NormalizedOrBottom Domain.NormalizedMap variable
             normalizedOrBottom = Ac.toNormalized patt
+
+data InKeys term =
+    InKeys
+        { symbol :: !Symbol
+        , keyTerm, mapTerm :: !term
+        }
+
+matchInKeys :: TermLike variable -> Maybe (InKeys (TermLike variable))
+matchInKeys (App_ symbol [keyTerm, mapTerm]) = do
+    hook2 <- (getHook . symbolHook) symbol
+    Monad.guard (hook2 == Map.in_keysKey)
+    return InKeys { symbol, keyTerm, mapTerm }
+matchInKeys _ = Nothing
```

Comment:
This is the primary use-case for the `Injection` class I recently added:

```suggestion
instance Injection (TermLike variable) (InKeys (TermLike variable)) where
    inject InKeys { symbol, keyTerm, mapTerm } =
        TermLike.mkApplySymbol symbol [keyTerm, mapTerm]
        
    retract termLike = do
        App_ symbol [keyTerm, mapTerm] <- termLike
        hook2 <- (getHook . symbolHook) symbol
        Monad.guard (hook2 == Map.in_keysKey)
        return InKeys { symbol, keyTerm, mapTerm }
        
matchInKeys :: TermLike variable -> Maybe (InKeys (TermLike variable))
matchInKeys = retract
```

The `inject` function will be useful below when you reconstruct the `in_keys` terms over the opaque Map items.